File: packages/go/examples/flights-gin/oidc_test.go
package main

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

// Test RSA key pair for signing OIDC tokens.
var testPrivateKey *rsa.PrivateKey
var testPublicJWK map[string]any

func init() {
	var err error
	testPrivateKey, err = rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		panic(err)
	}

	// Build JWK from public key.
	pub := testPrivateKey.PublicKey
	testPublicJWK = map[string]any{
		"kty": "RSA",
		"alg": "RS256",
		"use": "sig",
		"kid": "test-key-1",
		"n":   base64.RawURLEncoding.EncodeToString(pub.N.Bytes()),
		"e":   base64.RawURLEncoding.EncodeToString(big.NewInt(int64(pub.E)).Bytes()),
	}
}

// startJWKSServer starts a local HTTP server serving OIDC discovery + JWKS.
func startJWKSServer(t *testing.T) string {
	t.Helper()
	mux := http.NewServeMux()
	var serverURL string

	mux.HandleFunc("/.well-known/openid-configuration", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{
			"issuer":   serverURL,
			"jwks_uri": serverURL + "/jwks",
		})
	})
	mux.HandleFunc("/jwks", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{
			"keys": []any{testPublicJWK},
		})
	})

	ts := httptest.NewServer(mux)
	serverURL = ts.URL
	t.Cleanup(ts.Close)
	return ts.URL
}

// signTestToken creates a signed RS256 JWT.
func signTestToken(t *testing.T, issuer, audience string, claims map[string]any, expOffset time.Duration) string {
	t.Helper()

	header := map[string]any{"alg": "RS256", "typ": "JWT", "kid": "test-key-1"}
	headerJSON, _ := json.Marshal(header)
	headerB64 := base64.RawURLEncoding.EncodeToString(headerJSON)

	now := time.Now()
	payload := map[string]any{
		"iss": issuer,
		"aud": audience,
		"iat": now.Unix(),
		"exp": now.Add(expOffset).Unix(),
	}
	for k, v := range claims {
		payload[k] = v
	}
	payloadJSON, _ := json.Marshal(payload)
	payloadB64 := base64.RawURLEncoding.EncodeToString(payloadJSON)

	signingInput := headerB64 + "." + payloadB64
	hash := sha256.Sum256([]byte(signingInput))
	sig, err := rsa.SignPKCS1v15(rand.Reader, testPrivateKey, crypto.SHA256, hash[:])
	if err != nil {
		t.Fatal(err)
	}
	sigB64 := base64.RawURLEncoding.EncodeToString(sig)

	return headerB64 + "." + payloadB64 + "." + sigB64
}

func TestOIDC_ValidTokenWithEmail(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "test-audience", "")
	token := signTestToken(t, issuer, "test-audience", map[string]any{"email": "samir@example.com"}, time.Hour)
	principal, ok := v.Validate(token)
	if !ok {
		t.Fatal("expected validation to succeed")
	}
	if principal != "human:samir@example.com" {
		t.Fatalf("expected human:samir@example.com, got %s", principal)
	}
}

func TestOIDC_ValidTokenWithUsername(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "test-audience", "")
	token := signTestToken(t, issuer, "test-audience", map[string]any{"preferred_username": "samir", "sub": "u1"}, time.Hour)
	principal, ok := v.Validate(token)
	if !ok {
		t.Fatal("expected validation to succeed")
	}
	if principal != "human:samir" {
		t.Fatalf("expected human:samir, got %s", principal)
	}
}

func TestOIDC_ValidTokenSubOnly(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "test-audience", "")
	token := signTestToken(t, issuer, "test-audience", map[string]any{"sub": "service-xyz"}, time.Hour)
	principal, ok := v.Validate(token)
	if !ok {
		t.Fatal("expected validation to succeed")
	}
	if principal != "oidc:service-xyz" {
		t.Fatalf("expected oidc:service-xyz, got %s", principal)
	}
}

func TestOIDC_ExpiredToken(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "test-audience", "")
	token := signTestToken(t, issuer, "test-audience", map[string]any{"email": "x@x.com"}, -time.Hour)
	_, ok := v.Validate(token)
	if ok {
		t.Fatal("expected validation to fail for expired token")
	}
}

func TestOIDC_WrongIssuer(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator("https://wrong-issuer.example.com", "test-audience", fmt.Sprintf("%s/jwks", issuer))
	token := signTestToken(t, issuer, "test-audience", map[string]any{"email": "x@x.com"}, time.Hour)
	_, ok := v.Validate(token)
	if ok {
		t.Fatal("expected validation to fail for wrong issuer")
	}
}

func TestOIDC_WrongAudience(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "wrong-audience", "")
	token := signTestToken(t, issuer, "test-audience", map[string]any{"email": "x@x.com"}, time.Hour)
	_, ok := v.Validate(token)
	if ok {
		t.Fatal("expected validation to fail for wrong audience")
	}
}

func TestOIDC_InvalidSignature(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "test-audience", "")

	// Sign with a different key.
	otherKey, _ := rsa.GenerateKey(rand.Reader, 2048)
	header := map[string]any{"alg": "RS256", "typ": "JWT", "kid": "test-key-1"}
	headerJSON, _ := json.Marshal(header)
	headerB64 := base64.RawURLEncoding.EncodeToString(headerJSON)
	payload := map[string]any{
		"iss": issuer, "aud": "test-audience", "email": "x@x.com",
		"iat": time.Now().Unix(), "exp": time.Now().Add(time.Hour).Unix(),
	}
	payloadJSON, _ := json.Marshal(payload)
	payloadB64 := base64.RawURLEncoding.EncodeToString(payloadJSON)
	hash := sha256.Sum256([]byte(headerB64 + "." + payloadB64))
	sig, _ := rsa.SignPKCS1v15(rand.Reader, otherKey, crypto.SHA256, hash[:])
	token := headerB64 + "." + payloadB64 + "." + base64.RawURLEncoding.EncodeToString(sig)

	_, ok := v.Validate(token)
	if ok {
		t.Fatal("expected validation to fail for invalid signature")
	}
}

func TestOIDC_JWKSFetchFailure(t *testing.T) {
	v := NewOIDCValidator("http://localhost:1", "test-audience", "http://localhost:1/jwks")
	_, ok := v.Validate("not-a-jwt")
	if ok {
		t.Fatal("expected validation to fail")
	}
}

func TestOIDC_NonJWTString(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "test-audience", "")
	_, ok := v.Validate("not-a-jwt")
	if ok {
		t.Fatal("expected validation to fail for non-JWT")
	}
}

func TestOIDC_ExplicitJWKSURL(t *testing.T) {
	issuer := startJWKSServer(t)
	v := NewOIDCValidator(issuer, "test-audience", issuer+"/jwks")
	token := signTestToken(t, issuer, "test-audience", map[string]any{"email": "direct@example.com"}, time.Hour)
	principal, ok := v.Validate(token)
	if !ok {
		t.Fatal("expected validation to succeed with explicit JWKS URL")
	}
	if principal != "human:direct@example.com" {
		t.Fatalf("expected human:direct@example.com, got %s", principal)
	}
}

func TestOIDC_KeyRotationRefresh(t *testing.T) {
	// Generate a second key pair.
	rotatedKey, _ := rsa.GenerateKey(rand.Reader, 2048)
	rotatedPub := rotatedKey.PublicKey
	rotatedJWK := map[string]any{
		"kty": "RSA", "alg": "RS256", "use": "sig", "kid": "rotated-key-2",
		"n": base64.RawURLEncoding.EncodeToString(rotatedPub.N.Bytes()),
		"e": base64.RawURLEncoding.EncodeToString(big.NewInt(int64(rotatedPub.E)).Bytes()),
	}

	// Server that serves both keys.
	mux := http.NewServeMux()
	var serverURL string
	mux.HandleFunc("/.well-known/openid-configuration", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{"issuer": serverURL, "jwks_uri": serverURL + "/jwks"})
	})
	mux.HandleFunc("/jwks", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{"keys": []any{testPublicJWK, rotatedJWK}})
	})
	ts := httptest.NewServer(mux)
	serverURL = ts.URL
	t.Cleanup(ts.Close)

	v := NewOIDCValidator(serverURL, "test-audience", "")

	// First call with original key — caches JWKS.
	token1 := signTestToken(t, serverURL, "test-audience", map[string]any{"email": "first@example.com"}, time.Hour)
	p1, ok := v.Validate(token1)
	if !ok || p1 != "human:first@example.com" {
		t.Fatalf("first validation failed: %s %v", p1, ok)
	}

	// Second call with rotated kid — triggers JWKS refresh.
	header := map[string]any{"alg": "RS256", "typ": "JWT", "kid": "rotated-key-2"}
	headerJSON, _ := json.Marshal(header)
	headerB64 := base64.RawURLEncoding.EncodeToString(headerJSON)
	payload := map[string]any{
		"iss": serverURL, "aud": "test-audience", "email": "rotated@example.com",
		"iat": time.Now().Unix(), "exp": time.Now().Add(time.Hour).Unix(),
	}
	payloadJSON, _ := json.Marshal(payload)
	payloadB64 := base64.RawURLEncoding.EncodeToString(payloadJSON)
	sigInput := headerB64 + "." + payloadB64
	hash := sha256.Sum256([]byte(sigInput))
	sig, _ := rsa.SignPKCS1v15(rand.Reader, rotatedKey, crypto.SHA256, hash[:])
	token2 := headerB64 + "." + payloadB64 + "." + base64.RawURLEncoding.EncodeToString(sig)

	p2, ok := v.Validate(token2)
	if !ok {
		t.Fatal("rotated key validation failed — JWKS refresh did not work")
	}
	if !strings.Contains(p2, "rotated@example.com") {
		t.Fatalf("expected rotated@example.com principal, got %s", p2)
	}
}

File: examples/anip-ts/tests/oidc.test.ts
import { describe, it, expect, beforeAll, afterAll } from "vitest";
import { createOidcValidator } from "../src/oidc.js";
import * as jose from "jose";
import { createServer, type Server } from "node:http";

// Generate a test RSA key pair for signing OIDC tokens
let privateKey: jose.KeyLike;
let publicJwk: jose.JWK;
let jwksServer: Server;
let jwksPort: number;

const AUDIENCE = "test-service";

beforeAll(async () => {
  // Generate RSA key pair
  const { privateKey: priv, publicKey: pub } = await jose.generateKeyPair("RS256");
  privateKey = priv;
  publicJwk = await jose.exportJWK(pub);
  publicJwk.kid = "test-key-1";
  publicJwk.alg = "RS256";
  publicJwk.use = "sig";

  // Start a local JWKS server
  jwksServer = createServer((req, res) => {
    if (req.url === "/.well-known/openid-configuration") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({
        issuer: `http://localhost:${jwksPort}`,
        jwks_uri: `http://localhost:${jwksPort}/jwks`,
      }));
    } else if (req.url === "/jwks") {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ keys: [publicJwk] }));
    } else {
      res.writeHead(404);
      res.end();
    }
  });

  await new Promise<void>((resolve) => {
    jwksServer.listen(0, () => {
      jwksPort = (jwksServer.address() as { port: number }).port;
      resolve();
    });
  });
});

afterAll(() => {
  jwksServer.close();
});

function issuer() {
  return `http://localhost:${jwksPort}`;
}

async function signToken(claims: Record<string, unknown>, opts?: { expiresIn?: string; kid?: string }) {
  return new jose.SignJWT(claims as jose.JWTPayload)
    .setProtectedHeader({ alg: "RS256", kid: opts?.kid ?? "test-key-1" })
    .setIssuer(issuer())
    .setAudience(AUDIENCE)
    .setIssuedAt()
    .setExpirationTime(opts?.expiresIn ?? "1h")
    .sign(privateKey);
}

describe("OIDC Validator", () => {
  it("validates token with email claim → human:{email}", async () => {
    const validate = createOidcValidator({ issuerUrl: issuer(), audience: AUDIENCE });
    const token = await signToken({ email: "samir@example.com" });
    const principal = await validate(token);
    expect(principal).toBe("human:samir@example.com");
  });

  it("validates token with preferred_username → human:{username}", async () => {
    const validate = createOidcValidator({ issuerUrl: issuer(), audience: AUDIENCE });
    const token = await signToken({ preferred_username: "samir", sub: "user-123" });
    const principal = await validate(token);
    expect(principal).toBe("human:samir");
  });

  it("validates token with only sub → oidc:{sub}", async () => {
    const validate = createOidcValidator({ issuerUrl: issuer(), audience: AUDIENCE });
    const token = await signToken({ sub: "service-account-xyz" });
    const principal = await validate(token);
    expect(principal).toBe("oidc:service-account-xyz");
  });

  it("rejects expired token", async () => {
    const validate = createOidcValidator({ issuerUrl: issuer(), audience: AUDIENCE });
    const token = await signToken({ email: "samir@example.com" }, { expiresIn: "-1h" });
    const principal = await validate(token);
    expect(principal).toBeNull();
  });

  it("rejects wrong issuer", async () => {
    const validate = createOidcValidator({ issuerUrl: "https://wrong-issuer.example.com", audience: AUDIENCE });
    const token = await signToken({ email: "samir@example.com" });
    const principal = await validate(token);
    expect(principal).toBeNull();
  });

  it("rejects wrong audience", async () => {
    const validate = createOidcValidator({ issuerUrl: issuer(), audience: "wrong-audience" });
    const token = await signToken({ email: "samir@example.com" });
    const principal = await validate(token);
    expect(principal).toBeNull();
  });

  it("rejects invalid signature", async () => {
    const validate = createOidcValidator({ issuerUrl: issuer(), audience: AUDIENCE });
    // Sign with a different key
    const { privateKey: otherKey } = await jose.generateKeyPair("RS256");
    const token = await new jose.SignJWT({ email: "samir@example.com" })
      .setProtectedHeader({ alg: "RS256", kid: "test-key-1" })
      .setIssuer(issuer())
      .setAudience(AUDIENCE)
      .setIssuedAt()
      .setExpirationTime("1h")
      .sign(otherKey);
    const principal = await validate(token);
    expect(principal).toBeNull();
  });

  it("handles JWKS fetch failure gracefully", async () => {
    const validate = createOidcValidator({
      issuerUrl: "http://localhost:1", // unreachable
      audience: AUDIENCE,
      jwksUrl: "http://localhost:1/jwks",
    });
    const token = await signToken({ email: "samir@example.com" });
    const principal = await validate(token);
    expect(principal).toBeNull();
  });

  it("handles unknown kid by refreshing JWKS", async () => {
    // Generate a second key pair and add it to the server's key set mid-test
    const { privateKey: newPriv, publicKey: newPub } = await jose.generateKeyPair("RS256");
    const newJwk = await jose.exportJWK(newPub);
    newJwk.kid = "rotated-key-2";
    newJwk.alg = "RS256";
    newJwk.use = "sig";

    // Start a fresh server that serves both keys (simulating key rotation)
    const rotatedServer = createServer((req, res) => {
      if (req.url === "/.well-known/openid-configuration") {
        const port = (rotatedServer.address() as { port: number }).port;
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ issuer: `http://localhost:${port}`, jwks_uri: `http://localhost:${port}/jwks` }));
      } else if (req.url === "/jwks") {
        res.writeHead(200, { "Content-Type": "application/json" });
        res.end(JSON.stringify({ keys: [publicJwk, newJwk] }));
      } else {
        res.writeHead(404); res.end();
      }
    });
    const rotatedPort = await new Promise<number>((resolve) => {
      rotatedServer.listen(0, () => resolve((rotatedServer.address() as { port: number }).port));
    });

    try {
      const validate = createOidcValidator({
        issuerUrl: `http://localhost:${rotatedPort}`,
        audience: AUDIENCE,
      });

      // First call with original key — caches JWKS
      const token1Fixed = await new jose.SignJWT({ email: "first@example.com" } as jose.JWTPayload)
        .setProtectedHeader({ alg: "RS256", kid: "test-key-1" })
        .setIssuer(`http://localhost:${rotatedPort}`)
        .setAudience(AUDIENCE)
        .setIssuedAt()
        .setExpirationTime("1h")
        .sign(privateKey);
      expect(await validate(token1Fixed)).toBe("human:first@example.com");

      // Second call with rotated kid — triggers JWKS refresh
      const token2 = await new jose.SignJWT({ email: "rotated@example.com" } as jose.JWTPayload)
        .setProtectedHeader({ alg: "RS256", kid: "rotated-key-2" })
        .setIssuer(`http://localhost:${rotatedPort}`)
        .setAudience(AUDIENCE)
        .setIssuedAt()
        .setExpirationTime("1h")
        .sign(newPriv);
      const principal = await validate(token2);
      expect(principal).toBe("human:rotated@example.com");
    } finally {
      rotatedServer.close();
    }
  });

  it("returns null for non-JWT strings", async () => {
    const validate = createOidcValidator({ issuerUrl: issuer(), audience: AUDIENCE });
    const principal = await validate("not-a-jwt");
    expect(principal).toBeNull();
  });

  it("uses explicit jwksUrl when provided", async () => {
    const validate = createOidcValidator({
      issuerUrl: issuer(),
      audience: AUDIENCE,
      jwksUrl: `http://localhost:${jwksPort}/jwks`,
    });
    const token = await signToken({ email: "direct@example.com" });
    const principal = await validate(token);
    expect(principal).toBe("human:direct@example.com");
  });
});
